### DISCUSSION TITLE
Running a macOS App as a Firebase Server in lieu of a Swift Admin SDK

### DISCUSSION THREAD

**Discussion by bdrelling** [firebase/firebase-ios-sdk#8893] [Q&A] [answered]
As the title states, I'm looking for alternatives to the Node JS Admin SDK. I currently use Firestore and implemented Firebase Functions to bridge the gap for things like running tasks and sending messages to users. However, I'm getting tired of battling with the terribly worded errors and GCP IAM permission nonsense, among _many_ other issues, of Firebase Functions.

Before abandoning Firebase for my project entirely, I'm considering attempting to build a macOS process (or app, if necessary) that can run on a server and handle the functionality I'm currently using the Admin SDK for the following, and I believe I can replicate the behaviors I need with minimal risk/impact.

- Firestore -- I think I can do anything in the Client SDK that I could with the Admin SDK, the downside being I'd have to authenticate as a global admin, which introduces a security risk. Is there a way around this? Either way, not a huge concern because this project doesn't need to be too secure.
- Cloud Functions Scheduler -- can just run tasks naively
- Cloud Functions Firestore Triggers -- already included in the client SDK, but with a read cost
- Cloud Messaging -- I believe client SDKs can send messages? If not, I'll drop Cloud Messaging, so not a big deal.

Given the above, I believe this is a workable solution if I'm willing to run the risk of the server being tied to a common user auth key and add additional Cloud Firestore reads that would have been free from Firebase Functions Firestore Triggers. Nothing the server does is critical, meaning if it dies during the day, nothing bad happens. I can re-enable and it will continue on like nothing happened.

I'm posting in the hopes others might have suggestions / additional thoughts. Does my plan seem reasonable? Whatever I come up with, I'll document and share so others can replicate.

I know the Firebase team talked about the potential for a Swift Admin SDK in the past, preferably that could tie into Vapor. I can't imagine it's worth Firebase adding the maintenance cost for a Swift Admin SDK, but I'd love more info if anyone has any there as well.

Thanks!

**Comment by maksymmalyhin** [comment, marked as answer]
@bdrelling It will be interesting to learn a bit more about your use case. Would you mind sharing more details? From your description it looks like you have a backend implemented with in Swift with [Vapor](https://github.com/vapor/vapor) (running on a macOS server?) and you would like to use Firestore, Firebase Could Functions and Cloud Messaging from it. Does it look accurate? If yes, here are my initial thoughts on using Firebase SDKs for Apple platforms for this:
- theoretically you should be able to integrate Firebase SDKs with your Swift based backend running in macOS environment using Swift Package Manager. I must admit that I haven't tried it, so if you encounter issues please let us know. 
- you will need to be mindful of Firestore SDK offline cache which is optimized for applications and may not perform well in server environment
- you can trigger a cloud function with Firebase Functions SDK
- you can send Cloud Messages via a Cloud Function (e.g. like described [here](https://firebase.google.com/docs/functions/use-cases#notify_users_when_something_interesting_happens))

As an alternative solution, you can use Firebase HTTP API to access Firestore, Functions and Cloud Messaging directly form your backend. In this case you will be able to use any backend stack, optimize implementation for your use cases, but it will be a bit more tedious to implement:
- [Firestore HTTP API](https://firebase.google.com/docs/firestore/use-rest-api)
- [Firebase Cloud Functions HTTP API](https://firebase.google.com/docs/functions/http-events)
- [Firebase Cloud Messaging HTTP API](https://firebase.google.com/docs/cloud-messaging/http-server-ref)

**Reply by bdrelling** [reply]
I've been trying to run Firestore in my macOS app for a while now and I'm having no luck. It doesn't seem like it can make any outgoing requests and I'm unable to debug further.

Just to clarify, is their macOS support for Firestore, Messaging, etc.? I thought I saw a [table](https://firebase.google.com/docs/ios/learn-more#firebase_library_support_by_platform) confirming it was, but the [release notes](https://firebase.google.com/support/release-notes/ios) has info that conflicts with data from the table, so now I don't know what to believe.

**Reply by maksymmalyhin** [reply]
Firestore and Messaging should work on macOS now. If you encounter any issue with them on macOS please file a bug report so we can help with the fix.

**Reply by bdrelling** [reply]
@maksymmalyhin I've been running into nothing but issues with Firebase on macOS. Auth and Firestore both return errors. I'd say I'm pretty well-versed in how Firebase has to be configured for iOS, but could be missing something for macOS if there are any special steps needed to get it working.

This is as much time as I could devote to producing an issue for the Firebase team: https://github.com/firebase/firebase-ios-sdk/issues/8939

**Comment by maksymmalyhin** [comment]
cc @lahirumaramba

**Comment by bdrelling** [comment]
@maksymmalyhin Thank you for the reply, this is incredibly helpful. I will 100% share anything I do here publicly for others to learn from / expand upon, and I'll try to signal to the Firebase team as well when I've made any headway.

Re: "integrating Firebase SDKs with your Swift based backend running in macOS environment using SPM" -- The issue is that Vapor is typically going to run on Linux, which doesn't have access to `Foundation` and `URLSession`, which was why I didn't even attempt this yet. However, I can quickly add the dependency and see what errors I get.

I'll report back this week when I give it a shot!

Thanks again. Really appreciate your comments.

**Reply by mortenbekditlevsen** [reply]
A lot of Firebase is written in Objective-C, and there's a lot of the objc runtime bridging to Swift that's not available on non-Apple platforms.
So even though some part of Foundation and a big chunk of the URLSession APIs are available in Swift on Linux, that's still not enough to make the Firebase iOS sdk run on Linux.

I am also a huge fan of the thought of running Firebase with server side Swift and I started porting the RTDB implementation from objc to Swift. It's a huge endeavor, but also a lot of fun.

Which Firebase components are you interested in using from Linux?

**Reply by bdrelling** [reply]
Ah, right, I always forget about the Obj-C too.

But yeah, either way I was thinking that using the official client apps + the API as needed to run a server process (or for my purposes, just an app I can run while I'm using the computer, but could eventually turn into a headless process and host if I need 24/7 uptime).

For this project, I am only using Firestore and Messaging. And it isn't so much that I need them from Linux, but just that I want to be able to write my "backend" Cloud Functions code in Swift.

**Reply by mortenbekditlevsen** [reply]
Really cool - I love to hear about others that want server side swift + firebase to become a reality! :-)
 
I think that the suggestion from @maksymmalyhin of making a thin 'wrapper API' in swift around the Firestore HTTP API might be a good way to go when you're mainly interested in Firestore.

**Reply by bdrelling** [reply]
Yeah, the only issue is that if I'm going to make a thin wrapper, I'd rather invest that time into just building a thin backend myself.

**Reply by mortenbekditlevsen** [reply]
Makes sense - but if the back end is just one of many clients, then the other clients could still benefit greatly from all the power that Firestore gives you.

**Comment by lahirumaramba** [comment]
Hi @bdrelling, Thank you for creating this issue! We understand that there is an interest to use Firebase Admin APIs/Firebase Cloud Functions with server-side Swift. We have been discussing this topic with our teams lately, however this is not something the Firebase Admin SDK team currently have plans to work in the near future. We will use this issue to track any progress.

As you have mentioned in your initial post, using the Firebase REST APIs in a Swift server environment is one of the main workarounds for now... Additionally, you might be able to use a script written in Node.js or Python (or other officially supported language of the Admin SDK) as a way to bridge the space, though I understand that it might not be sustainable given the complexity and requirements of your projects.

If you are comfortable sharing we would love to hear more use cases and examples where you would prefer server-side Swift + Firebase services (Admin SDK). Thanks again for your feedback. These are very valuable inputs for us to further improve our SDKs.

b/205177095
